python/lsst/jointcal/jointcal.py

```diff
@@ -554,31 +543,74 @@ def runQuantum(self, butlerQC, inputRefs, outputRefs):
+                             inputs['inputCamera'], "setWcs")
         if self.config.doPhotometry:
-            self._put_output(butlerQC, outputs.outputPhotoCalib, outputRefs.outputPhotoCalib)
+            self._put_output(butlerQC, outputs.outputPhotoCalib, outputRefs.outputPhotoCalib,
+                             inputs['inputCamera'], "setPhotoCalib")
 
-    def _put_output(self, butlerQC, outputs, outputRefs):
-        """Butler put the output datasets to their appropriate datarefs.
+    def _put_output(self, butlerQC, outputs, outputRefs, camera, setter):
+        """Persist the output datasets to their appropriate datarefs.
 
         Parameters
         ----------
         butlerQC : `ButlerQuantumContext`
             A butler which is specialized to operate in the context of a
-            `lsst.daf.butler.Quantum`.
+            `lsst.daf.butler.Quantum`; This is the input to `runQuantum`.
         outputs : `dict` [`tuple`, `lsst.afw.geom.SkyWcs`] or
                   `dict` [`tuple, `lsst.afw.image.PhotoCalib`]
-            The fitted objects to put.
+            The fitted objects to persist.
         outputRefs : `list` [`OutputQuantizedConnection`]
-            The DatasetRefs to put the data to.
+            The DatasetRefs to persist the data to.
+        camera : `lsst.afw.cameraGeom.Camera`
+            The camera for this instrument, to get detector ids from.
+        setter : `str`
+            The method to call on the ExposureCatalog to set each output.
         """
+        schema = lsst.afw.table.ExposureTable.makeMinimalSchema()
+        schema.addField('visit', type='I', doc='Visit number')
+
+        def new_catalog(visit, size):
+            """Return an catalog ready to be filled with appropriate output."""
+            catalog = lsst.afw.table.ExposureCatalog(schema)
+            catalog.resize(size)
+            catalog['visit'] = visit
+            metadata = lsst.daf.base.PropertyList()
+            metadata.add("COMMENT", "Catalog id is detector id, sorted.")
+            metadata.add("COMMENT", "Only detectors with data have entries.")
+            return catalog
+
+        # count how many detectors have output for each visit
+        detectors_per_visit = collections.defaultdict(int)
+        for key in outputs:
+            detectors_per_visit[key[0]] += 1
+
         for ref in outputRefs:
-            key = (ref.dataId['visit'], ref.dataId['detector'])
-            butlerQC.put(outputs[key], ref)
+            visit = ref.dataId['visit']
+            catalog = new_catalog(visit, detectors_per_visit[visit])
+            # Iterate over every detector and skip the ones we don't have output for.
+            i = 0
+            for detector in camera:
+                detectorId = detector.getId()
+                key = (ref.dataId['visit'], detectorId)
+                if key not in outputs:
+                    continue  # skip detectors we don't have output for
```

Comment:
Is this worth a `self.log.debug` message for detectors that are skipped?

Reply:
I was going to say no (since this is mostly for tests that have limited data), but debug is probably the right level for that and it could be useful. Good suggestion.